compio-buf/src/io_vec_buf.rs

```diff
@@ -0,0 +1,197 @@
+use std::ops::{Deref, DerefMut};
+
+use crate::{
+    t_alloc, IndexedIter, IoBuf, IoBufMut, IoSlice, IoSliceMut, OwnedIterator, SetBufInit,
+};
+
+/// A type that's either owned or borrowed. Like [`Cow`](std::rc::Cow) but
+/// without the requirement of [`ToOwned`].
+pub enum MaybeOwned<'a, T> {
+    /// Owned.
+    Owned(T),
+    /// Borrowed.
+    Borrowed(&'a T),
+}
+
+impl<T> Deref for MaybeOwned<'_, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            MaybeOwned::Owned(t) => t,
+            MaybeOwned::Borrowed(t) => t,
+        }
+    }
+}
+
+/// A type that's either owned or mutably borrowed .
+pub enum MaybeOwnedMut<'a, T> {
+    /// Owned.
+    Owned(T),
+    /// Mutably borrowed.
+    Borrowed(&'a mut T),
+}
+
+impl<T> Deref for MaybeOwnedMut<'_, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            MaybeOwnedMut::Owned(t) => t,
+            MaybeOwnedMut::Borrowed(t) => t,
+        }
+    }
+}
+
+impl<T> DerefMut for MaybeOwnedMut<'_, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        match self {
+            MaybeOwnedMut::Owned(t) => t,
+            MaybeOwnedMut::Borrowed(t) => t,
+        }
+    }
+}
+
+/// A trait for vectored buffers.
+pub trait IoVectoredBuf: 'static {
+    /// The buffer.
+    type Buf: IoBuf;
+
+    /// The owned iterator that wraps `Self`.
+    type OwnedIter: OwnedIterator<Inner = Self> + IoBuf;
+
+    /// Collected [`IoSlice`]s of the buffers.
+    ///
+    /// # Safety
+    ///
+    /// The return slice will not live longer than self.
+    /// It is static to provide convenience from writing self-referenced
+    /// structure.
+    unsafe fn io_slices(&self) -> Vec<IoSlice> {
+        self.iter_buf().map(|buf| buf.as_io_slice()).collect()
+    }
+
+    /// An iterator over the buffers that's either owned or borrowed with
+    /// [`MaybeOwned`].
+    fn iter_buf(&self) -> impl Iterator<Item = MaybeOwned<'_, Self::Buf>>;
```

Comment:
Most of the changes look good to me. However, I don't understand the need for `MaybeOwned`. It seems like `MaybeOwned::Owned` isn't used at all. What is it designed for?

Reply:
For `opendal` `Buffer`. As it need to return partial view of `Bytes`, but `&Bytes` won't work like slice, so the only way is to return a new owned `Bytes` with new range

Reply:
> As it needs to return partial view of `Bytes`, but `&Bytes` won't work like slice

Could you provide more details? I assume `Bytes` should have already implemented `IoBuf`, right?

Reply:
Oh, I see. You mean we might encounter cases where new `Bytes` are created during the iteration process.

Reply:
We're implementing `IoVectoredBuf` for `Buffer`, meaning we need to iterate all `Bytes` in the `Buffer`. But `Buffer` has internal pointer (range) that indicate the start point, which may points to the middle of a `Bytes` meaning we may need to cut the first `Bytes`.

Reply:
> Oh, I see. You mean we might encounter cases where new `Bytes` are created during the iteration process since `opendal::Buffer` implements `Iterator<Item=Bytes>` directly.

Also owned iterator need to replicate this logic in a owned fashion, and that's the main reason of this refactor PR.

Reply:
I see. Do you think it's better to implement `IntoIterator` for `Buffer` instead?

Reply:
> I see. Do you think it's better to implement `IntoIterator` for `Buffer` instead? Like you suggested before.

That will be helpful

Reply:
> That will be helpful

Can we remove `MaybeOwned` now that `opendal::Buffer` has changed? It seems out of place in `IoVectoredBuf`.

Reply:
I don't think so. We still need to return either a newly cut `Bytes` or a reference to current `Bytes`.

Reply:
> I see. Do you think it's better to implement `IntoIterator` for `Buffer` instead? Like you suggested before.

Ok I tried, we may need something like this:

```rust
struct Iter<'a> {
  buf: Buffer,
  _p: PhantomData<&'a ()>
}

impl<'a> Iterator for Iter<'a> {
  type Item = &'a [u8];
  fn next(&mut self) -> Option<Self::Item> {
    // obtain a slice with unsafe, necessary for the same reason as std::slice::Iter
    // safe because of 'a here
    // then update internal cursor
  }
}
```

We are essentially suffer from lack of lending iterator.

Reply:
Oh...

I don't want `fn iter(&self) -> Iter<'_>` for `Buffer` that iterates `&[u8]`. Can we implement `Iterator<Item=Bytes>` by transform `Buffer` into a `Vec<Bytes>` first?

Reply:
No. The problems is that without `MaybeOwned` we have to return a `Iterator<Item = &impl Buf>`, with the lifetime bind to `self`.

Reply:
Got it. Let's add MaybeOwned back..

Reply:
> Oh...
> 
> I don't want `fn iter(&self) -> Iter<'_>` for `Buffer` that iterates `&[u8]`.
Actually, is it possible to have that privately in opendal core? It doesn't have to be a public API.